src/item.cpp

```diff
@@ -103,69 +103,58 @@ item::item()
     type = nullitem();
 }
 
-item::item(const std::string new_type, int turn, bool rand)
+item::item( const std::string new_type, int turn, int qty )
 {
     type = find_type( new_type );
     bday = turn;
     corpse = type->id == "corpse" ? &mtype_id::NULL_ID.obj() : nullptr;
-    name = type_name(1);
-    const bool has_random_charges = rand && type->spawn && type->spawn->rand_charges.size() > 1;
-    if( has_random_charges ) {
-        const auto charge_roll = rng( 1, type->spawn->rand_charges.size() - 1 );
-        charges = rng( type->spawn->rand_charges[charge_roll - 1], type->spawn->rand_charges[charge_roll] );
+    name = type_name();
+
+    if( qty >= 0 ) {
+        charges = qty;
+    } else {
+        if( type->spawn && type->spawn->rand_charges.size() > 1 ) {
+            const auto charge_roll = rng( 1, type->spawn->rand_charges.size() - 1 );
+            charges = rng( type->spawn->rand_charges[charge_roll - 1], type->spawn->rand_charges[charge_roll] );
+        } else {
+            charges = type->charges_default();
+        }
     }
-    // TODO: some item types use the same member (e.g. charges) for different things. Handle or forbid this.
+
     if( type->gun ) {
-        set_var( "magazine_converted", true );
-        charges = 0;
-        for( auto &gm : type->gun->built_in_mods ){
-            if(type_is_defined( gm) ){
-                item temp( gm, turn, rand );
+        for( const auto &mod : type->gun->built_in_mods ){
+            if( type_is_defined( mod ) ) {
+                item temp( mod, turn, qty );
                 temp.item_tags.insert("IRREMOVABLE");
                 contents.push_back( temp );
             }
         }
-
-        for( auto &gm : type->gun->default_mods ){
-            if(type_is_defined( gm ) ){
-                contents.push_back( item( gm, turn, rand ) );
+        for( const auto &mod : type->gun->default_mods ) {
+            if( type_is_defined( mod ) ) {
```

Comment:
Completely unrelated. This check here should not be needed. It should be done in the consistency checks of the item types after they have been loaded. Here, we should simply assume all ids in the container `default_mods` are valid.